### PR TITLE
Episode 4 Muscle coupling

### PR DESCRIPTION
**Pull request by JaroCamphuijsen**
This PR adds content to episode 4 about the second stage of model coupling with MUSCLE3. We will investigate the second sub-model and write a configuration file to manage the model execution flow in the yMMSL language.

### CODE REVIEW

File: episodes/04-muscle-couple.md [hunk 1]
```diff
@@ -1,102 +1,338 @@
...
+```
+
+
+::: solution
+
+```yaml
+ymmsl_version: v0.1
+
+model:
+  name: reaction_diffusion_python
```
Comment (JaroCamphuijsen) [inline episodes/04-muscle-couple.md:106]:
```suggestion
  name: reaction_diffusion
```

File: episodes/04-muscle-couple.md [hunk 2]
```diff
@@ -1,102 +1,338 @@
...
+
+  components:
+    macro:
+      implementation: diffusion_python
```
Comment (JaroCamphuijsen) [inline episodes/04-muscle-couple.md]:
```suggestion
      implementation: diffusion
```

File: episodes/04-muscle-couple.md [hunk 3]
```diff
@@ -1,102 +1,338 @@
...
+      ports:
+        o_i: state_out
+        s: state_in
+
+    micro:
+      implementation: reaction_python
```
Comment (JaroCamphuijsen) [inline episodes/04-muscle-couple.md]:
```suggestion
      implementation: reaction
```

File: episodes/04-muscle-couple.md [hunk 4]
```diff
@@ -1,102 +1,338 @@
...
+```yaml
+resources:
+  macro:
+    threads: 1
+  micro:
+    threads: 1
+```
 
-## Figures
+Note that we specify resources for each component, not for each implementation.
```
Comment (JaroCamphuijsen) [inline episodes/04-muscle-couple.md:177]: We haven't configured the implementations yet. Should we do that as well?
```suggestion
Finally we need to tell MUSCLE where the implementations of the sub-models live for each of the two sub-models as follows:

```yaml
implementations:
  reaction:
    virtual_env: /path/to/my/env
    executable: python
    args: /path/to/my/reaction.py
```
Note that we specify resources for each component, not for each implementation.
```

Comment (JaroCamphuijsen) [reply episodes/04-muscle-couple.md:177]: hmm, that is messed up....

File: episodes/data/model_coupling/coupled_model.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+# Configure Python logging
+logging.basicConfig()
+logging.getLogger('yatiml').setLevel(logging.WARNING)
+
+# Set this to logging.DEBUG to get some more information
+logging.getLogger().setLevel(logging.INFO)
+
+# Load configuration
+configuration = ymmsl.load(Path('reaction_diffusion_solution.ymmsl'))
```
Comment (JaroCamphuijsen) [inline episodes/data/model_coupling/coupled_model.py]:
```suggestion
configuration = ymmsl.load(Path('reaction_diffusion.ymmsl'))
```

File: episodes/data/model_coupling/coupled_model.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import logging
+from pathlib import Path
+
+import libmuscle
+from libmuscle.runner import run_simulation
+import ymmsl
+
+
+from diffusion import diffusion
+from reaction_solution import reaction
```
Comment (JaroCamphuijsen) [inline episodes/data/model_coupling/coupled_model.py]:
```suggestion
from reaction import reaction
```

**Review by JaroCamphuijsen — COMMENTED** [review]
I found a couple of bugs still when running the coupled simulation. I suppose that it is fine for now and we probably won't even make it all the way to the end tomorrow. 
I find the two options for running currently also slightly confusing, and i'm not sure whether the running from python example is really easier.

File: learners/setup.md [hunk 7]
```diff
@@ -2,48 +2,32 @@
...
 
-::::::::::::::::::::::::::::::::::::::: discussion
+MUSCLE3 is on PyPI as an ordinary Python package, so it can be installed via Pip in the usual way. It’s normally a good idea to make a virtual environment (virtualenv), if you don’t yet have one:
 
-### Details
+```bash
+python3 -m venv muscle3_venv
+. muscle3_venv/bin/activate
+pip install -U pip setuptools wheel
+pip install muscle3
```
Comment (JaroCamphuijsen) [inline learners/setup.md:19]: Without matplotlib the coupled simulation crashes when we try to run it from the `coupled_model.py` file. 
```suggestion
pip install muscle3
pip install matplotlib
```

**Comment by LourensVeen** [comment]
Thanks, good points all. The manager is really the standard way to run simulations, the runner is here for teaching basically (it's also nice for running tests). It is easier if all the programming experience you have is typing Python into PyCharm on Windows, because you don't have to learn how to make an executable script, and you don't have to learn how to use the command line. At least, that was the idea.

**Comment by LourensVeen** [comment]
Some of your comments seem to have gone missing here, but the implementations section in the yMMSL file is only used by the manager, the runner uses the dictionary to map implementation names to Python functions. So we don't need to add them until when we use the manager to run.

**Pull request merged**